File: kafka-streams/src/test/groovy/io/micronaut/configuration/kafka/streams/wordcount/WordCountListener.java
package io.micronaut.configuration.kafka.streams.wordcount;

import io.micronaut.configuration.kafka.annotation.KafkaKey;
import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;

import java.util.Collections;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import static io.micronaut.configuration.kafka.annotation.OffsetReset.EARLIEST;

@KafkaListener(offsetReset = EARLIEST, groupId = "WordCountListener")
public class WordCountListener {

    private final Map<String, Long> wordCounts = new ConcurrentHashMap<>();

    @Topic(WordCountStream.OUTPUT)
    void count(@KafkaKey String word, long count) {
        System.out.println("word = " + word);
        wordCounts.put(word, count);
    }

    public long getCount(String word) {
        Long num = wordCounts.get(word);
        if (num != null) {
            return num;
        }
        return 0;
    }

    public Map<String, Long> getWordCounts() {
        return Collections.unmodifiableMap(wordCounts);
    }
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/streams/WordCountListener.py
# tag::imports[]
from typing import Annotated

from java.lang import Long
from micronaut.configuration.kafka.annotation import KafkaKey, KafkaListener, OffsetReset, Topic
from micronaut.context.annotation import Requires
# end::imports[]


@Requires(property="spec.name", value="WordCountStreamTest")
# tag::clazz[]
@KafkaListener(offsetReset=OffsetReset.EARLIEST, groupId="WordCountListener")
class WordCountListener:

    def __init__(self):
        self.word_counts: dict[str, int] = {}

    @Topic("streams-wordcount-output")
    def count(self, word: Annotated[str, KafkaKey], count: Long) -> None:
        self.word_counts[word] = int(count)

    def get_count(self, word: str) -> int:
        return self.word_counts.get(word, 0)

    def get_word_counts(self) -> dict[str, int]:
        return dict(self.word_counts)
# end::clazz[]
